### ISSUE TITLE
Include version/status metadata for each CV

### ISSUE THREAD

**Issue by durack1** [WCRP-CMIP/CMIP6_CVs#28]
The repository already explicitly captures versioning information with the SHA1 for each commit, however it would be useful to consider adding "versioning" or "status/publication status" attributes or metadata for each CV.

The following email thread outlines some comments:

```
On 7/22/16, 10:22 AM, "Karl Taylor" wrote:

Hi Mark,

thanks for getting back to me on this.  I agree that version and
possibly status would be good to record in the files.  I'll consult with
Paul Durack who constructed these json files as to best syntax, but if
you have a strong preference, please let me know.

I missed the last telco, and I'll try to check with Balaji about
scheduling the next one.

hope you had an enjoyable vacation.
best regards,
Karl

On 7/22/16 3:52 AM, Elkington, Mark wrote:
Hi Karl

Just back from vacation.

Grid labelling is OK with me.  Good to see the CV git site as well.   I support Martin's request to have some version information included in the CVs - but it occurs to me that some CVs are probably going to become 'final' while others remain 'draft' - so might also be worth including a "publication_status" attribute in the version information.

Regards

Mark

P.S when is the next WIP telecon
```

**Comment by durack1** [comment]
@dnadeau4 @taylor13 @martinjuckes pinging you here

**Comment by durack1** [comment]
This might be helpful http://stackoverflow.com/questions/1792838/how-do-i-enable-ident-string-for-git-repos/

**Comment by durack1** [comment]
@taylor13 @dnadeau4 just for reference this is what @martinjuckes has in `CMIP6_variable.json`:
https://github.com/WCRP-CMIP/CMIP6_CVs/blob/master/CMIP6_variable.json#L16681-LL16686
```
    "info":{
        "author":"Martin Juckes",
        "creation_date":"Fri Jul 29 17:26:36 2016",
        "source":"Data Request",
        "version":"01.beta.31.01"
    }
```

**Comment by durack1** [comment]
The following should be added to resolve this issue:
```
    "version":{
        "author":"durack1 (Paul J. Durack)", ; # github identifier
        "creation_date":"Thu Aug 25 xx:xx:xx 2016", ; # automated time
        "institution":"PCMDI",
        "release":"1.0.0" ; # Need to release 1.0.0
    }
```

**Comment by durack1** [comment]
Reopening - #40 was reverted

**Comment by durack1** [comment]
```
On 9/14/16, 3:34 AM, martin.juckes wrote:

Hi Karl, Paul,

I understand that it is not possible to put the current SHA1 in the file, but I don't think that putting a past SHA1 in the file is a bad idea. It is bound to cause confusion. If you really want to include it, it should at least be identified correctly, e.g. labelled as "previous_commit" rather than "commit". We should at least avoid putting information which is obviously wrong in the file.

regards,
Martin
```

**Comment by durack1** [comment]
Ok @taylor13 @dnadeau4 @martinjuckes @markelkington so to close this out, what we currently have:
```
    "version":{
        "author":"Paul J. Durack <durack1@llnl.gov>",
        "commit":"003273faecf49e4124dac9655ca1b25f8a706e9b",
        "creation_date":"Thu Sep 1 10:29:58 2016 -0700",
        "institution_id":"PCMDI",
        "latest_tag_point":"None",
        "note":"Update versioning info"
}
```
Should then become:
```
    "meta":{
        "author":"Paul J. Durack <durack1@llnl.gov>",
        "latest_commit":"003273faecf49e4124dac9655ca1b25f8a706e9b",
        "creation_date":"Thu Sep 1 10:29:58 2016 -0700",
        "institution_id":"PCMDI",
        "latest_tag_point":"None",
        "note":"Update versioning info"
}
```

**Comment by taylor13** [comment]
@martinjuckes @charliepascoe @durack1 @momipsl @markelkington I'm reopening this issue in an attempt to find a way to make sure that both ES-DOC and the CMIP6 data request remain  consistent with the CMIP6 CV's.  

The version of the *collection* of CV's is recorded in each CV as shown here: https://github.com/WCRP-CMIP/CMIP6_CVs/issues/28#issuecomment-247200409 .  From this information, however, it is not clear what has changed since the last version was released.  I'm not particularly interested in any of the currently recorded version_metadata.  I would prefer seeing:

CMIP6 CV release label:  something like "6.X.X.X"
CMIP6 CV release date:  (currently named "creation_date")
CV_file_modification_date:  The date the contents of this file were most recently modified (ignoring changes in release label and release date.

This information would make it easy for someone to determine whether they had the latest version of each file in the CV and which files were updated for the current release.

Please say what would make it easier for you to remain consistent with the latest version of the CV's.

thanks,
Karl

**Comment by durack1** [comment]
All for reference, we currently have:
```json
"version_metadata":{
   "author":"Paul J. Durack <durack1@llnl.gov>",
   "creation_date":"Wed Apr 26 09:19:24 2017 -0700",
   "institution_id":"PCMDI",
   "latest_tag_point":"3.2.3 (30; gf11d16d)",
   "note":"Revise source_id BESM-2-7",
   "previous_commit":"cf5676f5f6ee42efd211608647aad38324505dd6"
}
```
What @taylor13 is recommending is something more like:
```json
"version_metadata":{
   "author":"Paul J. Durack <durack1@llnl.gov>",
   "creation_date":"Wed Apr 26 09:19:24 2017 -0700",
   "modification_date":"Wed Feb 1 07:14:22 2017 -0700", ! Date when CV contents last changed
   "institution_id":"PCMDI",
   "latest_tag_point":"6.0.0 (30; gf11d16d)", ! Impose repo-internal 6.x.x version
   "note":"Revise source_id BESM-2-7",
   "previous_commit":"cf5676f5f6ee42efd211608647aad38324505dd6"
}
```
@martinjuckes @charliepascoe @durack1 @momipsl @markelkington @eguil @bnlawrence @matthew-mizielinski If there are any suggestions, please drop them down here

**Comment by markelkington** [comment]
Hi

It is a good idea to include a release tag for a specific version of the “collection” of CV documents – as it would be for the ES-DOC and DREQ releases (I think DREQ does this already).

However what we also need as a modelling group is a higher level release “tag” for the consistent collection of CV, ES-DOC (experiment) and DREQ.  I am not suggesting that they all be brought under the same revision control system at this point  - probably too late for that – but some sort of communication approach that allows modelling groups to understand a complete “release” of these three sets of artefacts.

In the ES-DOC meeting yesterday an information page was proposed.  This would contain information like:

Release 1:
                CVs:  tag point 6.0.0
                ES-DOC experiment definition:  revision 1234
                DREQ:  release 1.56
                These items have been checked for completeness and consistency for CMIP, ScenarioMIP, DAMIP

Release 2:
                CVs:  tag point 6.12.0
                ES-DOC experiment definition:  revision 1274
                DREQ:  release 1.64
                These items have been checked for completeness and consistency for CMIP, ScenarioMIP, DAMIP, CFMIP, DCPP, FAFMIP, LUMIP, RFMIP

etc.
Note that this approach allows partial releases with consistency for the ‘early’ runs.  I also know this is somewhat undermined by picontrol having to be right for all MIPs.

It would also be really useful if this page could contain links to full change records for each item.  We and other modelling groups are having to develop change checking scripts for each of these collections of data to enable us to know if a particular release requires us to make changes to our production/post-processing configuration.  It would be more efficient if we could get a feeling for the scope of the change without having to go through this exercise each time. For example, it typically takes us a couple of weeks elapsed time to get the changes from a DREQ release into our mappings configuration (model outputs->data requirement).  Clearly we can’t do that for every release – and we need to pick which release we apply to our internal system.

Regards

Mark

**Comment by momipsl** [comment]
If you take a peek at how the CMIP6 vocabs are rewritten using the pyessv library within the ES-DOC space then you will observe that:  1. Each term is persisted as an independent file.  2.  Each collection is persisted as a manifest file.  3.  Each term is associated with a governance status.  The point I wish to highlight is the lower level of file granularity within pyessv, and hence the higher visibility of versioning changes.  

It is a real shame that in the vocab space there is evident risk of fragmentation.  We should be building downstream tools upon a common bedrock.  A coding sprint on this subject would be a good idea, although I am not sure who could fund it.

**Comment by matthew-mizielinski** [comment]
A few thoughts:

Given the frequency at which updates are being made to the content of this repository to add new information, I could argue for setting up a composite tag containing both
 * a version number of the form 6.x.y describing when changes to the structure of the CVs are made (hopefully few of these) or bugs are fixed,
 * a date (format `YYYYMMDD` or similar) describing the last date at which content was added to the CV files within the repository.

Tags of this form, e.g. `6.0.0-20170602` could be added around once per month, or as required if users need to pick up significant changes more quickly.  Users should then be able to work out from the tag whether there are any bugs/significant changes to pick up, or if there is new information, or both.

As the tag becomes part of an URL that can be used to get to the raw text files (e.g. the `3.2.2` in [this url](https://raw.githubusercontent.com/WCRP-CMIP/CMIP6_CVs/3.2.2/CMIP6_activity_id.json), then we can programmatically access the CV elements for a particular tagged version (either via a cached version or by a direct `request` to github).

Should you wish to validate the contents of any given JSON file in order to identify changes, briefly alluded to above if I've read the conversation correctly, then [this code](https://gist.github.com/matthew-mizielinski/612d291be47969e36b80a19bf4c2da08) may be useful to insert a simple checksum (I've used this to ensure that input files have not been hacked by hand in an attempt to ensure an audit trail is kept).

**Comment by durack1** [comment]
Will need to close #28 when this issue is resolved

**Comment by durack1** [comment]
This format should be the target:
```json
"version_metadata":{
   "author":"Paul J. Durack <durack1@llnl.gov>",
   "CV_collection_version":"6.5.2.22", 
   "CV_collection_modified":"Wed Jul 26 09:19:24 2017 -0700",
   "activity_id_CV_modified":"Wed Feb 1 07:14:22 2017 -0700", #The name depends on the CV
   "activity_id_CV_note":"Added FAFMIP to list", #The name depends on the CV
   "previous_commit":"cf5676f5f6ee42efd211608647aad38324505dd6"
}
```

**Comment by taylor13** [comment]
I think the CV_collection_version should be initiated with the latest released identified as 6.3.x.x.  The 3 is consistent with the current version of CMOR 3.3.X.  Does this seem like a good idea?  The numbers won't necessarily remain in sync though.

**Comment by durack1** [comment]
The `6.2.6` of https://goo.gl/v1drZl seems like a more appropriate starting point to me. The version of CMOR doesn't have much to do with the infrastructure that it uses

**Comment by taylor13** [comment]
We should make sure we're consistent here with whatever we agree to in https://goo.gl/86C4ZB

**Comment by durack1** [comment]
@taylor13 these changes are now implemented and should work off the base version `6.2.1.0` from now on..
